kubernetes-client-api/src/test/java/io/fabric8/kubernetes/client/ConfigSourcePrecedenceTest.java

```diff
@@ -0,0 +1,220 @@
+        System.setProperty("kubenamespace", Utils.filePath(ConfigSourcePrecedenceTest.class.getResource("/test-namespace")));
+        Config config = new ConfigBuilder().build();
+
+        // When + Then
+        assertThat(config)
+            .hasFieldOrPropertyWithValue("masterUrl", "https://172.28.128.4:8443/")
+            .hasFieldOrPropertyWithValue("namespace", "testns")
+            .hasFieldOrPropertyWithValue("autoOAuthToken", "token");
+      } finally {
+        System.clearProperty("KUBERNETES_SERVICE_HOST");
+        System.clearProperty("KUBERNETES_SERVICE_PORT");
+        System.clearProperty("kubernetes.auth.serviceAccount.token");
+        System.clearProperty("kubenamespace");
+      }
+    }
+  }
+
+  @Nested
+  @DisplayName("In Cluster mounted ServiceAccount")
+  class MountedServiceAccountSource {
+    @BeforeEach
+    void setUp() {
+      System.setProperty("kubeconfig", "/dev/null");
+      System.setProperty("KUBERNETES_SERVICE_HOST", "10.96.0.1");
+      System.setProperty("KUBERNETES_SERVICE_PORT", "443");
+      System.setProperty("kubernetes.auth.serviceAccount.token",
+          Utils.filePath(ConfigSourcePrecedenceTest.class.getResource("/test-serviceaccount/token")));
+      System.setProperty("kubenamespace", Utils.filePath(ConfigSourcePrecedenceTest.class.getResource("/test-namespace")));
+    }
+
+    @Test
+    @DisplayName("no other source provided, use ServiceAccount attributes in Config")
+    void whenNoOtherSourceProvided_thenUseServiceAccount() {
+      // Given
+      Config config = new ConfigBuilder().build();
+
+      // When + Then
+      assertThat(config)
+          .hasFieldOrPropertyWithValue("masterUrl", "https://10.96.0.1:443/")
+          .hasFieldOrPropertyWithValue("namespace", "testnsfrompath")
+          .extracting(Config::getAutoOAuthToken)
+          .asString()
+          .contains("token-from-mounted-serviceaccount");
+    }
+
+    @Nested
+    @DisplayName("user configuration overrides ServiceAccount's Config attributes")
+    class UserConfigurationOverServiceAccount extends UserConfigurationViaConfigBuilder {
+    }
+
+    @Test
+    @DisplayName("System Properties configured, then give precedence to System Properties")
+    void whenSystemPropertiesUsedForSomeFields_thenSystemPropertiesGivenPrecedence() {
+      try {
+        System.setProperty("kubernetes.master", "https://user-configuration-override:8443");
+        System.setProperty("kubernetes.namespace", "namespace-overridden");
+        System.setProperty("kubernetes.auth.token", "token-overridden");
+        Config config = new ConfigBuilder().build();
+
+        // When + Then
+        assertThat(config)
+            .hasFieldOrPropertyWithValue("masterUrl", "https://user-configuration-override:8443/")
+            .hasFieldOrPropertyWithValue("namespace", "namespace-overridden")
+            .hasFieldOrPropertyWithValue("autoOAuthToken", "token-overridden");
```

Comment:
Can you also add an assertion for one of the configurations that's read from the initial source? -> precedence values preserved <-